-Ystatistics accepts a list of phases for which to print stats

lrytz opened the pull request:
Also cleaned up handling of defaults in MultiChoiceSetting.

lrytz committed:
-Ystatistics accepts a list of phases, cleanups in MultiChoiceSetting

MultiChoiceSetting and Xlint with its deprecated aliases is now a bit
simpler, but there's still room for improvement, as noted in comments.

lrytz commented:
review by @som-snytt

som-snytt commented:
They call it a can of worms, not only because they quickly wriggle out of control, but because you get a queasy feeling in your stomach just at the thought of cracking it open.

The reason for the side-effecting default function was for the semantics of Xlint, so `scalac -Xlint:-flag -Xlint` does not enable the flag.  So `-Xlint` does not mean `-Xlint:ALL_the_things!`

Also, if I turn off a setting, order of args doesn't matter.

```
apm@mara:~/tmp$ vi linted.scala
apm@mara:~/tmp$ scalac -Xlint linted.scala
linted.scala:6: warning: non-nullary method overrides nullary method
trait Bar extends Foo { override def f() = 43 }
                                     ^
one warning found
apm@mara:~/tmp$ scalac -Xlint:-nullary-override -Xlint linted.scala
apm@mara:~/tmp$ scalac -Xlint:nullary-override linted.scala
linted.scala:6: warning: non-nullary method overrides nullary method
trait Bar extends Foo { override def f() = 43 }
                                     ^
one warning found
apm@mara:~/tmp$ scalac -Xlint -Xlint:-nullary-override linted.scala
```

I haven't had time yet before walking to school and coffee to see if this PR intends that.

som-snytt commented:
Back from walk to school.

So, I see the unit test fails, which checks the semantics for Xlint as described.

My other idea (I recall) was to have the default function take the setting object, but that's six of one or half-dozen of the other.  The whole layering of the Settings API is, needless to say, difficult to negotiate and probably too complicated; a target for refactoring.  I remember I wanted MultiChoice (in some version) to filter on choices, but couldn't without going out of scope.

lrytz commented:
i have another patch coming up, so better wait a bit with the review

On Friday, August 22, 2014, som-snytt notifications@github.com wrote:

> Back from walk to school.
> 
> So, I see the unit test fails, which checks the semantics for Xlint as
> described.
> 
> My other idea (I recall) was to have the default function take the setting
> object, but that's six of one or half-dozen of the other. The whole
> layering of the Settings API is, needless to say, difficult to negotiate
> and probably too complicated; a target for refactoring. I remember I wanted
> MultiChoice (in some version) to filter on choices, but couldn't without
> going out of scope.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/scala/scala/pull/3939#issuecomment-53080309.

som-snytt commented:
OK.  Someone mentioned -Xexperimental could be multichoiced, so I'd be willing to do something over the weekend for usability of Multichoice or just review your PR.

That would be after tequila, the kind with the worm in the bottle.

lrytz commented:
It should follow the semantics now:

`-Xlint` is equivalent to `-Xlint:_`.

The `_` adds all settings, but weakly: it adds only those that don't exist yet, and later explicit settings will override those added by `_`.

Finally, when specifying the same argument multiple times (positively or negatively), the first one counts. So `-Xlint:-a,a` is `-a`, while `-Xlint:a,-a` is `a`.

See the tests, and let me know if you agree it's the right semantics.

Dankeschön!

som-snytt commented:
For the last case, explicit flags in conflict, I know the current resolution is to respect turning it on.

Possibly it should just be an error or warning that "flag is already set the other way."

EDIT: here's where I said that's what javac does, FWIW:

https://github.com/scala/scala/commit/f81ec8d1f6481ddacfb27e743c6c58961e765f0e

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -211,10 +211,7 @@ class MutableSettings(val errorFn: String => Unit)
-      helper: MultiChoiceSetting => String = _ => choices.mkString(f"$descr:%n", f"%n  ", f"%n")
-  ) =
-    add(new MultiChoiceSetting(name, helpArg, descr, choices, default, helper))
+  def MultiChoiceSetting(name: String, helpArg: String, descr: String, choices: List[String], descriptions: List[String], default: Option[List[String]] = None) = add(new MultiChoiceSetting(name, helpArg, descr, choices, descriptions, default))
```
I have a few pixels left on my monitor, or I could hook up a second monitor if this line needs to be longer.

som-snytt commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -17,143 +17,150 @@ trait Warnings {
+  val warnPackageObjectClasses   = warnFlag("package-object-classes",
+                                   "Class or object defined in package object")
+  val warnUnsoundMatch           = warnFlag("unsound-match",
+                                   "Pattern match may not be typesafe")
```
I noticed at some point that not all the help text strings have punctuation at the end.

som-snytt commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -17,143 +17,150 @@ trait Warnings {
   val lint = {
     val description = "Enable or disable specific warnings"
-    val choices     = (lintWarnings map (_.name)).sorted
+    val choices     = (lintWarnings map (s => (s.name, s.helpDescription))).sortBy(_._1)
```
sorted is ok for the tuple.

som-snytt commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -17,143 +17,150 @@ trait Warnings {
-      def helpline(n: String) = lintWarnings.find(_.name == n).map(w => f"  ${w.name}%-25s ${w.helpDescription}%n")
-      choices flatMap (helpline(_)) mkString (f"$description:%n", "", f"%n")
-    } withPostSetHook { x =>
-      val Neg = "-"
```
FTR, that's because it used to be "no".

som-snytt commented on src/compiler/scala/tools/nsc/util/StatisticsInfo.scala:
```diff
@@ -14,12 +14,10 @@ abstract class StatisticsInfo {
   val retainedByType = Statistics.newByClass("#retained tree nodes by type")(Statistics.newCounter(""))
 
-  def print(phase: Phase) = if (phasesShown contains phase.name) {
+  def print(phase: Phase) = if (settings.YstatisticsPhases.value contains phase.name) {
```
MultiStringSetting.contains

lrytz replied:
Agree.

BUT, I have to point that the two really have to be the same, everything else is confusing.

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -557,53 +554,116 @@ class MutableSettings(val errorFn: String => Unit)
-    def addAll(): Unit     = (default getOrElse adderAll)()
 
-    // the semantics is: s is enabled, i.e., either s or (_ but not -s)
-    override def contains(s: String) = isChoice(s) && (value contains s) || (sawAll && !(value contains s"-$s"))
```
Is it not necessary to refine contains? I don't even know yet.

lrytz replied:
See my comment just above. This is one of the points of this change, that `value` contains the enabled settings. `value` is a public method of `AbsSetting`, it should contain the right information.

som-snytt replied:
One could look at the API the other way around, that `value` is just "what the user supplied" and `contains` is the test method for if something is enabled. `PhasesSetting.contains` is the (maybe bad) precedent.  It would be nice if `PhasesSetting` used `_` and also that a `PhasesSetting` could be mixed-in with `Multichoice`, but that is out of scope.

I'll think about it during review, but it would be nice if the is-enabled test is fast.  And, whoops, I just went to check that lint warnings are just a flag check, and I see they've been upgraded to a look-up.  At that point, one inclines to an enum set.

lrytz replied:
Yes, good point. To make the test fast, I'll go back to your solution of having private BooleanSettings, and having them set as a post hook of lint, makes sense.

For `value.contains` / `contains`, I really think we should keep them consistent. If `value` is "what was supplied", should it contain duplicates, `_`? If not, how should we decide what to de-sugar, and what not?

`PhasesSetting` does something on its own, but as you say, it should be merged with MultiChoice.

I'll try another thing: The variable `v = Nil` is inherited from `MultiStringSetting`. If it wasn't, we could implement it with getters / setters, that may give enough flexibility to get it the way we want.

som-snytt replied:
Actually, guess what, I'd like to offer you a PR which makes MultiChoice take an Enumeration.  It's actually very simple, and book-keeping with ValueSets is easier than working with Strings.  It's much more natural.

You really motivated me to move the code a bit further. 

I'm hampered because of watching Woody Allen's Husbands and Wives on Hulu.  Kind of devastating.

Anyway, it's kind of funny:

```
  val YstatisticsPhases = {
    object phases extends Enumeration { val parser, typer, patmat, erasure, cleanup = Value }
    val description = "Print compiler statistics for specific phases"
    MultiChoiceSetting(
      name    = "-Ystatistics",
      helpArg = "phase",
      descr   = description,
      domain  = phases,
      default = Some(List("_"))
    ) withPostSetHook { _ => scala.reflect.internal.util.Statistics.enabled = true }
  }
```

I'll try to finish up and PR to your repo before I fall asleep.

I also like the idea of leveraging much-despised Enumeration.  It kind of shines in this application.

Similarly,

```
  object LintWarnings extends Enumeration {
    case class LintWarning(name: String, help: String, yAliased: Boolean = false) extends Val(name) with MultiChoice

    val AdaptedArgs            = LintWarning("adapted-args",              "Warn if an argument list is modified to match the receiver.",               true)
```

and

```
def warnAdaptedArgs            = lint contains AdaptedArgs
```

you get the idea.

som-snytt replied:
It's evolutionary, but unwinds a couple of historical artifacts, from when I extended MultiString.  First of all, there is no notion of "choosing", as if MultiChoice could have no choices!  That was a leftover.  So, it doesn't extend MultiString. Then it's obvious that it should take an enum and value is a ValueSet and not a List[String].

lrytz replied:
Allright, sounds good! Thanks for taking over :)

som-snytt replied:
You can tell me later if it looks good.  I think it looks good so far.  Hard to tell in the middle of Woody Allen.

lrytz replied:
Here's another one - feel free to ignore, I can just do it with my own accessors, and it probably just makes things unnecessarily complicated.

I currently need a generalization of `_`. For `-Yopt` there are individual flags (like `dead-code`), but also group flags, like `level:method`, which enable a subset of the individual flags. I'd like to specify that some options just enable others. `_` would be just a (customizable) default.

som-snytt replied:
That's a good note on generalizing underscore.  I think that's what setting.enabling is for, but I'll have to think about it.  I did fall asleep after all.  Some days are a bit too long, or maybe I mean too short.

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -557,53 +554,116 @@ class MutableSettings(val errorFn: String => Unit)
+    def help: String = {
+      val choiceLenght = choices.map(_.length).max + 1
+      val formatStr = s"  %-${choiceLenght}s %s"
+      choices.zipAll(descriptions, "", "").map({
```
Minor typo `choiceLenght` and extra parens around braces.

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -557,53 +554,116 @@ class MutableSettings(val errorFn: String => Unit)
+      val formatStr = s"  %-${choiceLenght}s %s"
+      choices.zipAll(descriptions, "", "").map({
+        case (arg, descr) => formatStr.format(arg, descr)
+      }).mkString(s"$descr\n", "\n", "")
```
mkString needs final newline?  I prefer `f"text%n"` for windows/cygwin users, to avoid mixed line endings.

I must have just guessed at a field width before.

lrytz replied:
Final newline is not needed, seems to be added by the caller.

som-snytt commented on test/junit/scala/tools/testing/AssertUtil.scala:
```diff
@@ -19,6 +19,8 @@ object AssertUtil {
           throw e
+        else return
     }
+    throw new AssertionError("Expression did not throw!")
```
That's funny, we were recently discussing here this idiom in junit.  I like to have the annotation on the test method in case I somehow fumble the exception, so I catch, check the message and rethrow, and never do the fallback fail.

Edit: I was only describing what I do in Java in the office, not talking about this class; though I did use it as an illustration in that discussion; serves me right for being a know-it-all.

som-snytt commented on test/junit/scala/tools/nsc/settings/SettingsTest.scala:
```diff
@@ -54,7 +54,23 @@ class SettingsTest {
+
+    // the first setting counts
+    assertFalse(check("-Xlint:-missing-interpolator,missing-interpolator")(_.warnMissingInterpolator))
+    assertFalse(check("-Xlint:-missing-interpolator", "-Xlint:missing-interpolator")(_.warnMissingInterpolator))
```
I see that I started with the following test (which makes no sense), then ultimately consulted javac.

```
+ // two lint settings are first come etc; unlike -Y
+ assertTrue(check("-Xlint", "-Xlint:-warn-missing-interpolator")(_.warnMissingInterpolator))
```

lrytz replied:
OK, will align with javac for these cases.

som-snytt commented on src/compiler/scala/tools/nsc/settings/Warnings.scala:
```diff
@@ -17,143 +17,150 @@ trait Warnings {
+    else
+      nonLintWarnings ::= r
+    r
+  }
```
OK, let's see, before, the deprecated warning cascaded to the lint flag; now, the lint flag consults the deprecated alias.  

There are different ways of encoding it, but personally I prefer two methods, lintFlag and warnFlag, to the boolean param which does either a or b, and by the way don't supply an alias if not lint.

lrytz replied:
OK, will create two methods. I merged them because they were similar but slightly different :)

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -557,53 +554,116 @@ class MutableSettings(val errorFn: String => Unit)
-    private val adderAll = () => sawAll = true
-    private val noargs   = () => errorFn(s"'$name' requires an option. See '$name:help'.")
+
+    private val addedByUnderscore = collection.mutable.Set.empty[String]
```
This is the only part I don't like.  The previous incarnation of just setting a flag on underscore is simpler. 

On reflection, I'm not sure 50 LOC to do the book-keeping adds value (pun alert, maybe).

lrytz replied:
Not a beauty, I agree. But as I said, I'd really like to keep this `value` list consistent, it's part of the API of the setting. And it moves the complexity to the right point: when a new setting is parsed, we decide wether or not to add it to the list. Afterwards, querying for enabled options is straightforward.

lrytz replied:
I realize we can achieve the same with the `sawAll` flag that you introduced, and overwriting the `value` accessor. Will do that.

lrytz commented:
Thanks a lot for the review. It's indeed a bit of a worm-can. Anyway, I hope all is OK now.

lrytz commented:
Hah, some failed test because of the tightened `assertThrows`. Will take a look later.

lrytz committed:
Fix assertThrows, and the behaviors that it shadowed

The test in StdNamesTest was introduced in cff8b569, where newTermName
would throw on a negative length. In b090f97 this was changed to fix
the negative length, but the test was not adapted (as it didn't fail).

lrytz commented:
Finally tamed the tests.

@som-snytt, please let me know if you're OK taking another look at the worms. I squashed the changes, because I reverted some of the things you flagged, so it's easier to see the final diff.

som-snytt commented:
Yeah, sure. Better you than me. Then maybe multichoice experimental would be an easy exercise.

som-snytt committed:
Use Enumeration for MultiChoiceSetting

This is pretty easy, since a ValueSet is a BitSet.

When the setting is updated, recompute the current
set of values, which is cheap and succinct.

Checking a flag is also easy and fast.

Choices in MultiChoiceSettings may enable other choices.

som-snytt commented:
I couldn't see how to PR against lrytz in the github UI, so there's

https://github.com/scala/scala/pull/3945

which uses enums instead of List[String].

lrytz commented:
@som-snytt I closed #3945 and re-opened this one. I left your commit as-is for now, so it's easier to see the diff. I'll squash the two before merging, to ensure a green build.

Let me know what you think.

lrytz commented:
Need to rebase on top of `2.11.x`

lrytz commented:
@skyluc is looking into the IDE failure.

som-snytt commented:
That's interesting, your code looks the same as my edit yesterday, but maybe the MultiChoiceEnum helps typecheck.  Clever and brave to push the enum thing to the next level.

Is there a use case yet for expanding option? Or build it and they will come?  Nice idea. (Edit: OK, I just saw your other PR with the usage!)

lrytz committed:
Address PR feedback, fix MultiChoiceSetting.contains

lrytz commented:
`[locker.compiler] error: No space left on device` :-1:

lrytz commented:
PLS REBUILD/pr-scala@a7cd75d

scala-jenkins commented:
(kitty-note-to-self: ignore 54263546)
:cat: Roger! Rebuilding pr-scala for a7cd75dc. :rotating_light:

skyluc commented:
I found the reason for the failing tests in the Scala IDE build. It is a bit convoluted.
The problem related to this PR is that the current Scala IDE code support `MultiChoiceSetting` through `MultiStringSetting`. Before this PR, `MultiChoiceSetting` was extending `MultiStringSetting`.

So we have a `MatchError` in [some code](https://github.com/scala-ide/scala-ide/blob/4.0.0-m3/org.scala-ide.sdt.core/src/org/scalaide/ui/internal/preferences/ScalaCompilerPreferenceInitializer.scala#L24), which shortcut a piece of code initializing default values of other preferences, which disable part of the classpath validation => tests fail.

I can add support for `MultiChoiceSetting`, but the change in this PR is not source compatible (with vs without type parameter).
So, I'm not sure what to do.

lrytz commented:
Thanks @skyluc for figuring it out! How about adding `case _ => ""` temporarily, and then when this PR is merged, you can move to `case MultiChoiceSetting[_]`.

It would have been helpful to see the MatchErrorr stack trace in the jenkins log...

skyluc commented:
Yes, the match Error stack trace would have help me too. I'm getting a one liner when running it from Scala IDE, but I cannot find it in the Jenkins logs. It is part of the few small fixes I have to do after having looking at this code.

skyluc commented:
I don't think the catchall will work, it is likely to totally disable -Xlint, or to not take in account the user configuration.
I'll have to test, but I think the best is to remove the special support for this option for a while. It would still be configurable manually.

som-snytt commented:
"Ideally," MutableSettings#Setting or no wait AbsSettings#AbsSetting should declare `defaults` along with `choices`, as `List[String]` and defined as applied left to right, so normally single-valued settings have `defaults.size == 1`.  That's only a bit weird for IntSetting, which currently has an Int-valued default (which IDE must .toString for it to be useful anyway).  Then IDE has a uniform way of asking for defaults.

It's more natural to have a T-valued default, but it's not obvious that it is more useful.

I don't know whether this would break even more tools.

lrytz commented:
I thought the codebase uses what I initially had, but then I checked the [style guide](http://docs.scala-lang.org/style/scaladoc.html#general_style), which kind of surprised me, but since I don't care much, so I went there.

The problem with the one you propose is that IntelliJ doesn't support it, while it supports the other two. When writing your proposed style, I end up with the following when hitting return after `Aba`

```
  /** Aba
    * 
   * 
   */
  def foo
```

som-snytt commented:
Yeah, the old style guide was formed before there was any consensus.  It's like a meteor impact from the early days of the solar system.  It didn't kill any life back then, but today, people keep falling into the crater.

https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/Global.scala#L270

@retronym must have an opinion because he's an IntelliJ guy, right?

https://github.com/scala/scala/blame/2.11.x/src/reflect/scala/reflect/internal/tpe/TypeMaps.scala#L211

Or @soc also

https://github.com/scala/scala/blame/2.11.x/src/library/scala/Boolean.scala#L23

Paul's first commit was "correct":

https://github.com/scala/scala/blob/305f49ce8f7358636bf81a7aca29d8ab42d98ed4/src/library/scala/Boolean.scala

Not that I care that much; but I can't access our corp network due to SSO outage, so here I sit.  Actually, I do care, since switched from tabs to spaces, I mean from old alignment to new.

lrytz commented:
I went back to old-style comments.

@skyluc can you do the necessary change to the IDE, or should I refactor the patch to avoid the type parameter?

skyluc commented:
I'm working on pushing a change to allow yours. But I have problems with our own PR validator, I'm trying to see what is happening.

lrytz commented:
OK, thanks for the heads up!

skyluc commented:
We merged the change in Scala IDE, the tests should pass now.

lrytz commented:
PLS REBUILD/pr-scala@e6e638b

scala-jenkins commented:
(kitty-note-to-self: ignore 54604188)
:cat: Roger! Rebuilding pr-scala for e6e638bd. :rotating_light:

lrytz commented:
@skyluc that worked, thanks!
@som-snytt if the patch looks good to you now, I'll squash my commit into yours to get all builds green.

som-snytt commented:
@lrytz Squash away.  Squash is nearly in season.  Thanks again for letting me participate.

lrytz commented:
I should either close the light or the windows, otherwise there'll be many mosquitoes to squash. Anyway, patch seems ready for review now.

One problem that I found: in `Global`, the field [`settings` is a `def`](https://github.com/scala/scala/blob/2.11.x/src/compiler/scala/tools/nsc/Global.scala#L79). So if you're coding somewhere in the compiler, this will give you a type error:

```
if (settings.myMultichoice contains settings.myMultichoiceValues.someValue)
```

Because `settings` is not a path. So you need to do

```
val s = settings
if (s.myMultichoice contains s.myMultichoiceValues.someValue)
```

som-snytt commented:
I was going to say, The only part I didn't crack open was `compute`, but then I looked at it on my lunch break.  I needed a white board to work it out.

```
//The previous expression:
def compute = yeas | (vetoable_expansions &~ nays)
def vetoable_expansions = if (all) values else empty
// now:
def compute = (yeas filter simple) | (vetoable_expansions &~ nays)
def vetoable_expansions = if (all) values filter simple else (yeas filterNot simple).expanded
```

I made a gist!

https://gist.github.com/som-snytt/45b82e58310c8ebb1e5f

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -553,57 +552,200 @@ class MutableSettings(val errorFn: String => Unit)
+                        else domain.ValueSet.empty
+
+      value = nonExpanding(yeas) | expandedYeas | expandedAll
+    }
```
It's only a dozen lines of code, but I know it will be touched again because we'll want to negate expanders, like `Xlint:_,-beginner` to turn off beginner advice.  Possibly the expression in the gist is more clear. (But same refactored logic.)

https://gist.github.com/som-snytt/45b82e58310c8ebb1e5f#file-expander-scala-L33

You can tell it felt complex because I used an underscore in a name.

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -553,57 +552,200 @@ class MutableSettings(val errorFn: String => Unit)
+      def expand(vs: domain.ValueSet): domain.ValueSet = vs flatMap {
+        // expand
+        case c @ ChoiceOrVal(_, _, others) if others.nonEmpty => expand(domain.ValueSet(others: _*))
+        case c @ ChoiceOrVal(_, _, _) if !nays(c)             => domain.ValueSet(c)
```
This filtering on nays can be deferred; simpler logic at the expense of the allocation, etc.

som-snytt commented on src/compiler/scala/tools/nsc/settings/MutableSettings.scala:
```diff
@@ -553,57 +552,200 @@ class MutableSettings(val errorFn: String => Unit)
+       */
+      def expand(vs: domain.ValueSet): domain.ValueSet = vs flatMap {
+        // expand
+        case c @ ChoiceOrVal(_, _, others) if others.nonEmpty => expand(domain.ValueSet(others: _*))
```
Deferring the filter on nays allows the Nil case as at

https://gist.github.com/som-snytt/45b82e58310c8ebb1e5f#file-expander-scala-L29

Fewer things to keep track of on my fingers using rubber bands and bits of twine.

lrytz commented:
```
> def compute = (yeas filter simple) | (vetoable_expansions &~ nays)
> def vetoable_expansions = if (all) values filter simple else (yeas filterNot simple).expanded
```

I should also use that whiteboard more often, much easier to understand.

> we'll want to negate expanders

What would be the best rules?
- `_,-ab,a` would only remove `b`
- `ab,-ab` should add both `a` and `b`
- `abc,-ab` should add all of `a`, `b`, `c`
- `abc,-a,-b` would only add `c`

som-snytt commented:
We'll leave it to tlc to work out what `tlc -Xlint:tlc,-tlc` means.

lrytz commented:
I'm sure you have an opinion, given that you [forked scala](https://github.com/som-snytt/scala) :-)

More seriously, I think the rules for negative expanding choices are quite filigrane, but I'm OK adding that. So basically, `_` is weaker negative expanding, positive expanding are stronger than negative expanding. The surprising one would be `:abc,-ab`, which would enable all `a`, `b`, `c`.

Or do you see a more intuitive solution?

som-snytt commented:
No, I think you were right not to overengineer it; it's simpler to disallow -ab for now.

lrytz committed:
Make MultiChoiceSetting.compute easier to understand

lrytz commented:
PTAL

som-snytt commented:
LGTM.  Even I can understand it.

If you want, you can remove the obsolete comments about nays on expand. And then if you want, you can have ChoiceOrVal.unapply take domain.Value, and remove the None case, with confidence.

lrytz commented:
@som-snytt thanks for the continuous reviews! Funny how long this PR grew.
